File: java/logbrew-java/src/test/java/co/logbrew/sdk/TelemetryContextTest.java
package co.logbrew.sdk;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicReference;

/** Dependency-free tests for shared Java telemetry context. */
public final class TelemetryContextTest {
    private int testsRun;

    public static void main(String[] args) {
        new TelemetryContextTest().run();
    }

    private void run() {
        testRuntimeDefaultsCoverEveryEventType();
        testClientAndEventContextsMergeAndDetachInputs();
        testRuntimeDefaultsCanBeDisabledWithoutDroppingExplicitContext();
        testContextValidationRejectsUnsafeOrAmbiguousValues();
        testTraceHelpersPromoteExactCorrelation();
        testRequestTelemetryPromotesExactContextAcrossSignals();
        testConcurrentContextCaptureRemainsBoundedAndDeterministic();
        System.out.println("java telemetry context tests ok (" + testsRun + " tests)");
    }

    private void testRuntimeDefaultsCoverEveryEventType() {
        LogBrewClient client = LogBrewClient.create("LOGBREW_API_KEY", "logbrew-java", "0.1.0");
        enqueueAll(client);

        String payload = client.previewJson();
        assertEquals(7, count(payload, "\"context\":"), "context count");
        assertEquals(7, count(payload, "\"schemaVersion\": 1"), "schema version count");
        assertEquals(7, count(payload, "\"name\": \"java\""), "runtime name count");
        assertContains(payload, "\"operatingSystem\"");
        assertContains(payload, "\"architecture\"");
        assertNotContains(payload, "LOGBREW_API_KEY");
        testsRun++;
    }

    @SuppressWarnings("unchecked")
    private void testClientAndEventContextsMergeAndDetachInputs() {
        Map<String, String> clientTags = new LinkedHashMap<>();
        clientTags.put("plan", "team");
        clientTags.put("region", "eu");
        TelemetryContext clientContext = TelemetryContext.builder()
            .resource(TelemetryResource.builder()
                .service("checkout-api", "1.4.0")
                .runtime("custom-java", "21")
                .device(null, "container", null)
                .build())
            .session("session_01", "session_00")
            .tags(clientTags)
            .build();

        Map<String, String> eventTags = new LinkedHashMap<>();
        eventTags.put("operation", "checkout");
        eventTags.put("plan", "enterprise");
        TelemetryContext eventContext = TelemetryContext.builder()
            .resource(TelemetryResource.builder()
                .service("checkout-api", "1.5.0")
                .device(null, null, "wasm32")
                .application("checkout-worker", null, "20260803.1")
                .build())
            .trace(
                "CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC",
                null,
                "DDDDDDDDDDDDDDDD",
                null
            )
            .subject("user_42", "user")
            .tags(eventTags)
            .build();

        clientTags.put("region", "mutated");
        eventTags.put("operation", "mutated");
        TelemetryContext merged = TelemetryContext.merge(clientContext, eventContext);
        Map<String, Object> value = merged.asMap();
        Map<String, Object> resource = (Map<String, Object>) value.get("resource");
        assertEquals(
            Map.of("name", "checkout-api", "version", "1.5.0"),
            resource.get("service"),
            "merged service"
        );
        assertEquals(
            Map.of("name", "custom-java", "version", "21"),
            resource.get("runtime"),
            "merged runtime"
        );
        assertEquals(
            Map.of("model", "container", "architecture", "wasm32"),
            resource.get("device"),
            "merged device"
        );
        assertEquals(
            Map.of("name", "checkout-worker", "build", "20260803.1"),
            resource.get("application"),
            "merged application"
        );
        assertEquals(
            Map.of("operation", "checkout", "plan", "enterprise", "region", "eu"),
            value.get("tags"),
            "merged tags"
        );
        assertEquals(
            Map.of("traceId", "cccccccccccccccccccccccccccccccc", "parentSpanId", "dddddddddddddddd"),
            value.get("trace"),
            "normalized trace"
        );

        LogBrewClient client = LogBrewClient.create(
            "LOGBREW_API_KEY",
            "logbrew-java",
            "0.1.0",
            LogBrewClientOptions.builder()
                .context(clientContext)
                .disableRuntimeContext(true)
                .build()
        );
        client.log(
            "evt_context",
            "2026-08-03T00:00:00Z",
            LogAttributes.create("checkout failed", "error").context(eventContext)
        );
        String payload = client.previewJson();
        assertContains(payload, "\"version\": \"1.5.0\"");
        assertContains(payload, "\"operation\": \"checkout\"");
        assertNotContains(payload, "mutated");
        testsRun++;
    }

    private void testRuntimeDefaultsCanBeDisabledWithoutDroppingExplicitContext() {
        LogBrewClient absent = LogBrewClient.create(
            "LOGBREW_API_KEY",
            "logbrew-java",
            "0.1.0",
            LogBrewClientOptions.builder().disableRuntimeContext(true).build()
        );
        absent.log(
            "evt_absent",
            "2026-08-03T00:00:00Z",
            LogAttributes.create("safe", "info")
        );
        assertNotContains(absent.previewJson(), "\"context\"");

        LogBrewClient explicit = LogBrewClient.create(
            "LOGBREW_API_KEY",
            "logbrew-java",
            "0.1.0",
            LogBrewClientOptions.builder()
                .disableRuntimeContext(true)
                .context(TelemetryContext.builder().tag("plan", "team").build())
                .build()
        );
        explicit.log(
            "evt_explicit",
            "2026-08-03T00:00:00Z",
            LogAttributes.create("safe", "info")
        );
        String payload = explicit.previewJson();
        assertContains(payload, "\"context\"");
        assertContains(payload, "\"plan\": \"team\"");
        assertNotContains(payload, "\"resource\"");
        testsRun++;
    }

    private void testContextValidationRejectsUnsafeOrAmbiguousValues() {
        assertError(
            () -> TelemetryContext.builder().build(),
            "must include resource, trace, session, subject, or tags"
        );
        assertError(
            () -> TelemetryContext.builder()
                .trace("00000000000000000000000000000000", null, null, null)
                .build(),
            "traceId must be 32 non-zero hex characters"
        );
        assertError(
            () -> TelemetryContext.builder().session("same", "same").build(),
            "previousId must differ from id"
        );
        assertError(
            () -> TelemetryContext.builder().subject("subject_1", "person").build(),
            "kind must be anonymous or user"
        );
        assertError(
            () -> TelemetryContext.builder().tag("bad key", "safe").build(),
            "tag key is invalid"
        );
        assertError(
            () -> TelemetryContext.builder().tag("plan", "safe\u0085unsafe").build(),
            "tag value for plan is invalid"
        );
        TelemetryContext.Builder tooManyTags = TelemetryContext.builder();
        for (int index = 0; index < 33; index++) {
            tooManyTags.tag("tag." + index, "safe");
        }
        assertError(tooManyTags::build, "tags must contain 1-32 entries");
        TelemetryContext.Builder maximumClientTags = TelemetryContext.builder();
        for (int index = 0; index < 32; index++) {
            maximumClientTags.tag("client." + index, "safe");
        }
        TelemetryContext fullClientContext = maximumClientTags.build();
        TelemetryContext extraEventTag = TelemetryContext.builder().tag("event.extra", "safe").build();
        assertError(
            () -> TelemetryContext.merge(fullClientContext, extraEventTag),
            "tags must contain 1-32 entries"
        );
        testsRun++;
    }

    @SuppressWarnings("unchecked")
    private void testTraceHelpersPromoteExactCorrelation() {
        LogBrewTraceContext trace = LogBrewTraceContext.create(
            "4BF92F3577B34DA6A3CE929D0E0E4736",
            "B7AD6B7169203331",
            "00F067AA0BA902B7",
            "01"
        );
        TelemetryContext shared = TelemetryContext.builder().tag("operation", "checkout").build();
        TelemetryContext traced = LogBrewTrace.contextWithTrace(trace, shared);
        Map<String, Object> traceValue = (Map<String, Object>) traced.asMap().get("trace");
        assertEquals(trace.traceId(), traceValue.get("traceId"), "trace id");
        assertEquals(trace.spanId(), traceValue.get("spanId"), "span id");
        assertEquals(trace.parentSpanId(), traceValue.get("parentSpanId"), "parent span id");
        assertEquals(Boolean.TRUE, traceValue.get("sampled"), "sampled flag");

        LogBrewClient client = LogBrewClient.create(
            "LOGBREW_API_KEY",
            "logbrew-java",
            "0.1.0",
            LogBrewClientOptions.builder().disableRuntimeContext(true).build()
        );
        client.log(
            "evt_traced",
            "2026-08-03T00:00:00Z",
            LogAttributes.create("checkout failed", "error")
                .metadata(LogBrewTrace.metadataWithTrace(trace, Map.of("stage", "payment")))
                .context(traced)
        );
        String payload = client.previewJson();
        assertContains(payload, "\"traceId\": \"4bf92f3577b34da6a3ce929d0e0e4736\"");
        assertContains(payload, "\"spanId\": \"b7ad6b7169203331\"");
        assertContains(payload, "\"sampled\": true");
        assertContains(payload, "\"operation\": \"checkout\"");
        testsRun++;
    }

    private void testRequestTelemetryPromotesExactContextAcrossSignals() {
        LogBrewClient client = LogBrewClient.create(
            "LOGBREW_API_KEY",
            "logbrew-java",
            "0.1.0",
            LogBrewClientOptions.builder().disableRuntimeContext(true).build()
        );
        LogBrewTraceContext trace = LogBrewTraceContext.create(
            "4bf92f3577b34da6a3ce929d0e0e4736",
            "b7ad6b7169203331",
            "00f067aa0ba902b7",
            "01"
        );
        LogBrewHttpRequestTelemetry telemetry = LogBrewHttpRequestTelemetry.start(
            client,
            "POST",
            "/checkout/:cart_id",
            trace
        );
        telemetry.finishSpanAndMetric(
            "evt_request_span",
            "evt_request_metric",
            "2026-08-03T00:00:00Z",
            202,
            42.0
        );

        String payload = client.previewJson();
        assertEquals(2, count(payload, "\"context\":"), "request context count");
        assertEquals(2, count(payload, "\"schemaVersion\": 1"), "request schema count");
        assertEquals(2, count(payload, "\"sampled\": true"), "request sampled count");
        assertContains(payload, "\"type\": \"span\"");
        assertContains(payload, "\"type\": \"metric\"");
        testsRun++;
    }

    private void testConcurrentContextCaptureRemainsBoundedAndDeterministic() {
        TelemetryContext clientContext = TelemetryContext.builder().tag("region", "eu").build();
        TelemetryContext eventContext = TelemetryContext.builder().tag("operation", "checkout").build();
        LogBrewClient client = LogBrewClient.create(
            "LOGBREW_API_KEY",
            "logbrew-java",
            "0.1.0",
            LogBrewClientOptions.builder()
                .context(clientContext)
                .disableRuntimeContext(true)
                .build()
        );
        AtomicReference<Throwable> failure = new AtomicReference<>();
        List<Thread> workers = new ArrayList<>();
        for (int worker = 0; worker < 8; worker++) {
            int workerId = worker;
            Thread thread = new Thread(() -> {
                try {
                    for (int event = 0; event < 100; event++) {
                        client.log(
                            "evt_context_" + workerId + "_" + event,
                            "2026-08-03T00:00:00Z",
                            LogAttributes.create("checkout", "info").context(eventContext)
                        );
                    }
                } catch (Throwable error) {
                    failure.compareAndSet(null, error);
                }
            }, "logbrew-context-" + workerId);
            workers.add(thread);
            thread.start();
        }
        for (Thread worker : workers) {
            try {
                worker.join();
            } catch (InterruptedException error) {
                Thread.currentThread().interrupt();
                throw new AssertionError("context worker interrupted", error);
            }
        }
        if (failure.get() != null) {
            throw new AssertionError("concurrent context capture failed", failure.get());
        }

        String payload = client.previewJson();
        assertEquals(800, client.pendingEvents(), "concurrent event count");
        assertEquals(0, client.droppedEvents(), "concurrent dropped count");
        assertEquals(800, count(payload, "\"context\":"), "concurrent context count");
        assertEquals(800, count(payload, "\"region\": \"eu\""), "concurrent client tag count");
        assertEquals(800, count(payload, "\"operation\": \"checkout\""), "concurrent event tag count");
        testsRun++;
    }

    private static void enqueueAll(LogBrewClient client) {
        client.release(
            "evt_release",
            "2026-08-03T00:00:00Z",
            ReleaseAttributes.create("1.2.3")
        );
        client.environment(
            "evt_environment",
            "2026-08-03T00:00:01Z",
            EnvironmentAttributes.create("production")
        );
        client.issue(
            "evt_issue",
            "2026-08-03T00:00:02Z",
            IssueAttributes.create("Checkout failed", "error")
        );
        client.log(
            "evt_log",
            "2026-08-03T00:00:03Z",
            LogAttributes.create("Checkout failed", "error")
        );
        client.span(
            "evt_span",
            "2026-08-03T00:00:04Z",
            SpanAttributes.create(
                "checkout",
                "4bf92f3577b34da6a3ce929d0e0e4736",
                "b7ad6b7169203331",
                "error"
            )
        );
        client.action(
            "evt_action",
            "2026-08-03T00:00:05Z",
            ActionAttributes.create("checkout.submit", "failure")
        );
        client.metric(
            "evt_metric",
            "2026-08-03T00:00:06Z",
            MetricAttributes.create("checkout.duration", "histogram", 42.0, "ms", "delta")
        );
    }

    private static int count(String value, String needle) {
        int total = 0;
        int cursor = 0;
        while ((cursor = value.indexOf(needle, cursor)) >= 0) {
            total++;
            cursor += needle.length();
        }
        return total;
    }

    private static void assertError(Runnable callback, String message) {
        try {
            callback.run();
        } catch (SdkException error) {
            assertContains(error.getMessage(), message);
            return;
        }
        throw new AssertionError("expected SdkException containing " + message);
    }

    private static void assertContains(String value, String needle) {
        if (!value.contains(needle)) {
            throw new AssertionError("expected " + value + " to contain " + needle);
        }
    }

    private static void assertNotContains(String value, String needle) {
        if (value.contains(needle)) {
            throw new AssertionError("expected " + value + " to omit " + needle);
        }
    }

    private static void assertEquals(int expected, int actual, String label) {
        if (expected != actual) {
            throw new AssertionError(label + ": expected " + expected + " but got " + actual);
        }
    }

    private static void assertEquals(Object expected, Object actual, String label) {
        if (!expected.equals(actual)) {
            throw new AssertionError(label + ": expected " + expected + " but got " + actual);
        }
    }
}

File: go/logbrew/telemetry_context_test.go
package logbrew

import (
	"encoding/json"
	"fmt"
	"reflect"
	"runtime"
	"strings"
	"testing"
)

func TestRuntimeContextDefaultsEveryEventType(t *testing.T) {
	client := sampleClient(t)
	enqueueAll(t, client)
	if err := client.Metric("evt_metric_001", "2026-06-02T10:00:06Z", MetricAttributes{
		Name: "checkout.requests", Kind: "counter", Value: 1, Unit: "{request}", Temporality: "delta",
	}); err != nil {
		t.Fatal(err)
	}

	var payload struct {
		Events []struct {
			Type       string         `json:"type"`
			Attributes map[string]any `json:"attributes"`
		} `json:"events"`
	}
	preview, err := client.PreviewJSON()
	if err != nil {
		t.Fatal(err)
	}
	if err := json.Unmarshal([]byte(preview), &payload); err != nil {
		t.Fatal(err)
	}

	expected := map[string]any{
		"schemaVersion": float64(1),
		"resource": map[string]any{
			"runtime":         map[string]any{"name": "go", "version": runtime.Version()},
			"operatingSystem": map[string]any{"name": runtime.GOOS},
			"device":          map[string]any{"architecture": runtime.GOARCH},
		},
	}
	if len(payload.Events) != 7 {
		t.Fatalf("expected seven event types, got %d", len(payload.Events))
	}
	for _, event := range payload.Events {
		if !reflect.DeepEqual(event.Attributes["context"], expected) {
			t.Fatalf("%s context mismatch: got %#v want %#v", event.Type, event.Attributes["context"], expected)
		}
	}
	if strings.Contains(preview, "LOGBREW_SERVER_API_KEY") || strings.Contains(preview, "LOGBREW_API_KEY") {
		t.Fatalf("runtime context leaked configuration values: %s", preview)
	}
}

func TestClientAndEventContextsMergeAndDetachInputs(t *testing.T) {
	clientContext := &TelemetryContext{
		SchemaVersion: 1,
		Resource: &TelemetryResource{
			Service: &TelemetryNamedVersion{Name: "checkout-api", Version: "1.4.0"},
			Runtime: &TelemetryNamedVersion{Name: "custom-go", Version: "1.24.0"},
			Device:  &TelemetryDevice{Model: "container"},
		},
		Trace: &TelemetryTraceContext{
			TraceID: "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA",
			SpanID:  "BBBBBBBBBBBBBBBB",
			Sampled: boolPtr(true),
		},
		Session: &TelemetrySessionContext{ID: "session_01", PreviousID: "session_00"},
		Tags:    map[string]string{"plan": "team", "region": "eu"},
	}
	eventContext := &TelemetryContext{
		SchemaVersion: 1,
		Resource: &TelemetryResource{
			Service:     &TelemetryNamedVersion{Name: "checkout-api", Version: "1.5.0"},
			Device:      &TelemetryDevice{Architecture: "wasm32"},
			Application: &TelemetryApplication{Name: "checkout-worker", Build: "20260803.1"},
		},
		Trace: &TelemetryTraceContext{
			TraceID:      "CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC",
			ParentSpanID: "DDDDDDDDDDDDDDDD",
		},
		Subject: &TelemetrySubjectContext{ID: "user_42", Kind: "user"},
		Tags:    map[string]string{"feature": "one-click", "plan": "enterprise"},
	}
	client, err := NewClient(Config{
		APIKey:                "LOGBREW_API_KEY",
		SDKName:               "logbrew-go",
		SDKVersion:            "0.1.0",
		Context:               clientContext,
		DisableRuntimeContext: true,
	})
	if err != nil {
		t.Fatal(err)
	}
	if err := client.Log("evt_context", "2026-08-03T00:00:00Z", LogAttributes{
		Message: "checkout failed", Level: "error", Context: eventContext,
	}); err != nil {
		t.Fatal(err)
	}

	clientContext.Resource.Service.Version = "mutated"
	clientContext.Tags["region"] = "mutated"
	eventContext.Resource.Application.Name = "mutated"
	eventContext.Tags["feature"] = "mutated"

	context := previewEventContext(t, client)
	expected := map[string]any{
		"schemaVersion": float64(1),
		"resource": map[string]any{
			"service":     map[string]any{"name": "checkout-api", "version": "1.5.0"},
			"runtime":     map[string]any{"name": "custom-go", "version": "1.24.0"},
			"device":      map[string]any{"model": "container", "architecture": "wasm32"},
			"application": map[string]any{"name": "checkout-worker", "build": "20260803.1"},
		},
		"trace": map[string]any{
			"traceId":      "cccccccccccccccccccccccccccccccc",
			"parentSpanId": "dddddddddddddddd",
		},
		"session": map[string]any{"id": "session_01", "previousId": "session_00"},
		"subject": map[string]any{"id": "user_42", "kind": "user"},
		"tags": map[string]any{
			"feature": "one-click", "plan": "enterprise", "region": "eu",
		},
	}
	if !reflect.DeepEqual(context, expected) {
		t.Fatalf("merged context mismatch: got %#v want %#v", context, expected)
	}
}

func TestPublicTelemetryContextHelpersNormalizeMergeAndDetach(t *testing.T) {
	base := &TelemetryContext{
		SchemaVersion: 1,
		Resource: &TelemetryResource{
			Service: &TelemetryNamedVersion{Name: " checkout-api ", Version: "1.4.0"},
		},
		Tags: map[string]string{"plan": " team "},
	}
	normalized, err := ValidateTelemetryContext(base)
	if err != nil {
		t.Fatal(err)
	}
	base.Resource.Service.Name = "mutated"
	base.Tags["plan"] = "mutated"
	if normalized.Resource.Service.Name != "checkout-api" || normalized.Tags["plan"] != "team" {
		t.Fatalf("public validation did not normalize and detach: %#v", normalized)
	}

	override := &TelemetryContext{
		SchemaVersion: 1,
		Resource: &TelemetryResource{
			Service: &TelemetryNamedVersion{Name: "checkout-api", Version: "1.5.0"},
		},
		Tags: map[string]string{"operation": "checkout"},
	}
	merged, err := MergeTelemetryContexts(normalized, override)
	if err != nil {
		t.Fatal(err)
	}
	override.Resource.Service.Version = "mutated"
	override.Tags["operation"] = "mutated"
	if merged.Resource.Service.Version != "1.5.0" ||
		!reflect.DeepEqual(merged.Tags, map[string]string{"operation": "checkout", "plan": "team"}) {
		t.Fatalf("public merge did not apply field overrides and detach: %#v", merged)
	}
}

func TestRuntimeContextCanBeDisabledWithoutDroppingExplicitContext(t *testing.T) {
	explicit, err := NewClient(Config{
		APIKey:                "LOGBREW_API_KEY",
		SDKName:               "logbrew-go",
		SDKVersion:            "0.1.0",
		DisableRuntimeContext: true,
		Context: &TelemetryContext{
			SchemaVersion: 1,
			Tags:          map[string]string{"plan": "team"},
		},
	})
	if err != nil {
		t.Fatal(err)
	}
	if err := explicit.Log("evt_explicit", "2026-08-03T00:00:00Z", LogAttributes{Message: "safe", Level: "info"}); err != nil {
		t.Fatal(err)
	}
	if got := previewEventContext(t, explicit); !reflect.DeepEqual(got, map[string]any{
		"schemaVersion": float64(1), "tags": map[string]any{"plan": "team"},
	}) {
		t.Fatalf("unexpected explicit-only context: %#v", got)
	}

	absent, err := NewClient(Config{
		APIKey:                "LOGBREW_API_KEY",
		SDKName:               "logbrew-go",
		SDKVersion:            "0.1.0",
		DisableRuntimeContext: true,
	})
	if err != nil {
		t.Fatal(err)
	}
	if err := absent.Log("evt_absent", "2026-08-03T00:00:00Z", LogAttributes{Message: "safe", Level: "info"}); err != nil {
		t.Fatal(err)
	}
	var payload struct {
		Events []struct {
			Attributes map[string]any `json:"attributes"`
		} `json:"events"`
	}
	preview, err := absent.PreviewJSON()
	if err != nil {
		t.Fatal(err)
	}
	if err := json.Unmarshal([]byte(preview), &payload); err != nil {
		t.Fatal(err)
	}
	if _, ok := payload.Events[0].Attributes["context"]; ok {
		t.Fatalf("disabled client unexpectedly captured context: %s", preview)
	}
}

func TestTelemetryContextValidationRejectsUnsafeOrAmbiguousValues(t *testing.T) {
	invalid := []struct {
		name    string
		context *TelemetryContext
		message string
	}{
		{"schema version", &TelemetryContext{SchemaVersion: 2, Tags: map[string]string{"plan": "team"}}, "schemaVersion must be 1"},
		{"empty", &TelemetryContext{SchemaVersion: 1}, "must include resource, trace, session, subject, or tags"},
		{"runtime name", &TelemetryContext{SchemaVersion: 1, Resource: &TelemetryResource{Runtime: &TelemetryNamedVersion{Version: "1.24"}}}, "runtime name is required"},
		{"zero trace", &TelemetryContext{SchemaVersion: 1, Trace: &TelemetryTraceContext{TraceID: strings.Repeat("0", 32)}}, "traceId must be 32 non-zero hex characters"},
		{"same session", &TelemetryContext{SchemaVersion: 1, Session: &TelemetrySessionContext{ID: "same", PreviousID: "same"}}, "previousId must differ from id"},
		{"subject kind", &TelemetryContext{SchemaVersion: 1, Subject: &TelemetrySubjectContext{ID: "user_1", Kind: "person"}}, "kind must be anonymous or user"},
		{"tag key", &TelemetryContext{SchemaVersion: 1, Tags: map[string]string{"bad key": "value"}}, "tags key is invalid"},
		{"control text", &TelemetryContext{SchemaVersion: 1, Tags: map[string]string{"plan": "safe\u0085unsafe"}}, "tags value for plan is invalid"},
	}
	for _, test := range invalid {
		t.Run(test.name, func(t *testing.T) {
			_, err := NewClient(Config{
				APIKey: "LOGBREW_API_KEY", SDKName: "logbrew-go", SDKVersion: "0.1.0",
				DisableRuntimeContext: true, Context: test.context,
			})
			if err == nil || !strings.Contains(err.Error(), test.message) {
				t.Fatalf("unexpected error: %v", err)
			}
		})
	}

	baseTags := make(map[string]string, 32)
	for index := 0; index < 32; index++ {
		baseTags[fmt.Sprintf("base.%02d", index)] = "safe"
	}
	client, err := NewClient(Config{
		APIKey: "LOGBREW_API_KEY", SDKName: "logbrew-go", SDKVersion: "0.1.0",
		DisableRuntimeContext: true,
		Context:               &TelemetryContext{SchemaVersion: 1, Tags: baseTags},
	})
	if err != nil {
		t.Fatal(err)
	}
	err = client.Log("evt_too_many_tags", "2026-08-03T00:00:00Z", LogAttributes{
		Message: "safe", Level: "info",
		Context: &TelemetryContext{SchemaVersion: 1, Tags: map[string]string{"event.tag": "safe"}},
	})
	if err == nil || !strings.Contains(err.Error(), "merged telemetry context tags must contain 1-32 entries") {
		t.Fatalf("unexpected merged tag error: %v", err)
	}
	if client.PendingEvents() != 0 {
		t.Fatalf("invalid merged context entered queue")
	}
}

func TestTraceHelpersPromoteActiveTraceIntoFirstClassContext(t *testing.T) {
	trace := testTrace(t, "B7AD6B7169203331")
	ctx := ContextWithLogBrewTrace(t.Context(), trace)
	logAttributes := LogAttributesWithTrace(ctx, LogAttributes{
		Message: "checkout failed", Level: "error",
	})
	issueAttributes := IssueAttributesWithTrace(ctx, IssueAttributes{
		Title: "Checkout failed", Level: "error",
		Context: &TelemetryContext{SchemaVersion: 1, Tags: map[string]string{"operation": "checkout"}},
	})
	actionAttributes := ActionAttributesWithTrace(ctx, ActionAttributes{
		Name: "checkout.submitted",
	})
	metricAttributes := MetricAttributesWithTrace(ctx, MetricAttributes{
		Name: "checkout.duration", Kind: "histogram", Value: 42, Unit: "ms", Temporality: "delta",
	})
	spanAttributes, err := SpanAttributesFromTraceContext(TraceContextSpanInput{
		Trace: trace, Name: "checkout", Status: "error",
	})
	if err != nil {
		t.Fatal(err)
	}
	for name, context := range map[string]*TelemetryContext{
		"log":    logAttributes.Context,
		"issue":  issueAttributes.Context,
		"action": actionAttributes.Context,
		"metric": metricAttributes.Context,
		"span":   spanAttributes.Context,
	} {
		if context == nil || context.Trace == nil {
			t.Fatalf("%s trace was not promoted into typed context: %#v", name, context)
		}
		if context.Trace.TraceID != trace.TraceID ||
			context.Trace.SpanID != trace.SpanID ||
			context.Trace.ParentSpanID != trace.ParentSpanID ||
			context.Trace.Sampled == nil || !*context.Trace.Sampled {
			t.Fatalf("unexpected %s trace context: %#v", name, context.Trace)
		}
	}
	if issueAttributes.Context.Tags["operation"] != "checkout" {
		t.Fatalf("trace promotion dropped event context: %#v", issueAttributes.Context)
	}

	client, err := NewClient(Config{
		APIKey: "LOGBREW_API_KEY", SDKName: "logbrew-go", SDKVersion: "0.1.0",
		DisableRuntimeContext: true,
		Context: &TelemetryContext{
			SchemaVersion: 1,
			Tags:          map[string]string{"plan": "team"},
		},
	})
	if err != nil {
		t.Fatal(err)
	}
	if err := client.Issue("evt_traced", "2026-08-03T00:00:00Z", issueAttributes); err != nil {
		t.Fatal(err)
	}
	context := previewEventContext(t, client)
	if !reflect.DeepEqual(context["trace"], map[string]any{
		"traceId":      trace.TraceID,
		"spanId":       trace.SpanID,
		"parentSpanId": trace.ParentSpanID,
		"sampled":      true,
	}) {
		t.Fatalf("queued issue missing exact span correlation: %#v", context)
	}
	if !reflect.DeepEqual(context["tags"], map[string]any{"operation": "checkout", "plan": "team"}) {
		t.Fatalf("queued issue missing merged tags: %#v", context)
	}
}

func previewEventContext(t *testing.T, client *Client) map[string]any {
	t.Helper()
	var payload struct {
		Events []struct {
			Attributes map[string]any `json:"attributes"`
		} `json:"events"`
	}
	preview, err := client.PreviewJSON()
	if err != nil {
		t.Fatal(err)
	}
	if err := json.Unmarshal([]byte(preview), &payload); err != nil {
		t.Fatal(err)
	}
	context, ok := payload.Events[0].Attributes["context"].(map[string]any)
	if !ok {
		t.Fatalf("event context missing: %s", preview)
	}
	return context
}
